torchtitan/components/checkpoint.py

```diff
@@ -357,117 +569,12 @@ def _async_with_pinned_memory(self, checkpoint_id: str) -> None:
             self.staging = True
             self.staging_id = checkpoint_id
 
-    def save(self, curr_step: int, force: bool = False) -> None:
-        """
-        force = True will force the checkpoint to be saved, even if the interval
-        has not been reached.
-        This only happens when train_state.step == job_config.training.steps, or
-        for initial seed checkpoint.
-        """
-        if not self._should_save(curr_step, force):
-            return
-
-        begin = time.monotonic()
-        checkpoint_id = self._create_checkpoint_id(curr_step)
-        self._async_wait()
-        # This GC is called for async checkpoint as it is useless to do
-        # GC right after async_save -- the CPU memory is not able to be
-        # freed until _async_wait()
-        if force:
-            self._save_last_step(curr_step)
-        elif self.async_mode == AsyncMode.ASYNC_WITH_PINNED_MEM:
-            GarbageCollection.collect("GC collection invoked by checkpointer.")
-            self._async_with_pinned_memory(checkpoint_id)
-        elif self.async_mode == AsyncMode.ASYNC:
-            GarbageCollection.collect("GC collection invoked by checkpointer.")
-            self.async_future = dcp.async_save(
-                self.states, checkpoint_id=checkpoint_id, process_group=self.pg
-            )
-        else:
-            save_with_gc(self.states, checkpoint_id=checkpoint_id)
-        self.reset()
-        self._purge_stale_checkpoints()
-
-        logger.info(
-            "Finished saving the checkpoint (or staging if async is enabled)"
-            f"in {time.monotonic() - begin:.2f} seconds."
-        )
-
-    def maybe_wait_for_staging(self) -> None:
+    def _purge_stale_checkpoints(self):
         if (
-            self.enable_checkpoint
-            and self.async_mode == AsyncMode.ASYNC_WITH_PINNED_MEM
-            and self.staging
+            self.keep_latest_k > 0
+            and dist.get_rank() == 0
```

Comment:
For my learning purpose, why do we only do the purge for `dist.get_rank() == 0`?

Reply:
We assume that all ranks can access the same files. That's the assumption of DCP. If we let all ranks to purge, then the file systems will complain.